File: java-adaptor/src/main/java/org/judal/storage/java/MapRecord.java
package org.judal.storage.java;

/*
 * © Copyright 2016 the original author.
 * This file is licensed under the Apache License version 2.0.
 * You may not use this file except in compliance with the license.
 * You may obtain a copy of the License at:
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.
 */

import java.io.Serializable;

/*
 * © Copyright 2016 the original author.
 * This file is licensed under the Apache License version 2.0.
 * You may not use this file except in compliance with the license.
 * You may obtain a copy of the License at:
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.
 */

import java.sql.Types;
import java.util.Map;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Set;
import java.util.TreeSet;

import javax.jdo.JDOException;
import javax.jdo.JDOUserException;

import org.judal.metadata.TableDef;
import org.judal.metadata.ViewDef;
import org.judal.metadata.ColumnDef;
import org.judal.serialization.BytesConverter;
import org.judal.storage.ConstraintsChecker;
import org.judal.storage.FieldHelper;
import org.judal.storage.table.TableDataSource;
import org.judal.storage.table.impl.AbstractRecord;
import org.judal.storage.java.internal.CaseInsensitiveValuesMap;

import com.knowgate.debug.DebugFile;

/**
 * Implementation of Record interface using an hash map to hold column values
 * @author Sergio Montoro Ten
 * @version 1.0
 */
public class MapRecord extends AbstractRecord implements JavaRecord {

	private static final long serialVersionUID = 10000l;

	private CaseInsensitiveValuesMap values;
	private HashMap<Integer, String> columnIndexes;

	/**
	 * Constructor
	 */
	public MapRecord(ViewDef tableDefinition) {
		this(tableDefinition,null,null);
	}

	/**
	 * Constructor
	 */
	public MapRecord(ViewDef tableDefinition, FieldHelper fieldHelper) {
		this(tableDefinition, fieldHelper, null);
	}

	/**
	 * Constructor
	 */
	public MapRecord(ViewDef tableDefinition, ConstraintsChecker constraintsChecker) {
		this(tableDefinition, null, constraintsChecker);
	}
	
	/**
	 * Constructor
	 */
	public MapRecord(ViewDef tableDefinition, FieldHelper fieldHelper, ConstraintsChecker constraintsChecker) {
		super(tableDefinition, fieldHelper, constraintsChecker);
		values = new CaseInsensitiveValuesMap();
		columnIndexes = new HashMap<Integer, String>();
		int i = 0;
		for (ColumnDef cdef: tableDefinition.getColumns())
			columnIndexes.put(++i, cdef.getName());
	}

	/**
	 * Constructor
	 * @param dataSource TableDataSource
	 * @param tableName String
	 * @throws JDOException 
	 * @throws JDOUserException if table or view is not found at DataSource metadata
	 */
	public MapRecord(TableDataSource dataSource, String tableName) throws JDOException {
		this(getViewDefForName(dataSource, tableName), dataSource.getFieldHelper());	
	}

	/**
	 * Constructor
	 * @param dataSource TableDataSource
	 * @param tableName String
	 * @param fieldHelper FieldHelper
	 * @throws JDOException 
	 * @throws JDOUserException if table or view is not found at DataSource metadata
	 */
	public MapRecord(TableDataSource dataSource, String tableName, FieldHelper fieldHelper) throws JDOException {
		this(getViewDefForName(dataSource, tableName), fieldHelper);
	}

	/**
	 * Constructor
	 * @param dataSource TableDataSource
	 * @param tableName String
	 * @param constraintsChecker ConstraintsChecker
	 * @throws JDOException 
	 * @throws JDOUserException if table or view is not found at DataSource metadata
	 */
	public MapRecord(TableDataSource dataSource, String tableName, ConstraintsChecker constraintsChecker) throws JDOException {
		this(dataSource, tableName, dataSource.getFieldHelper(), constraintsChecker);
	}
	
	/**
	 * Constructor
	 * @param dataSource TableDataSource
	 * @param tableName String
	 * @param fieldHelper FieldHelper
	 * @param constraintsChecker ConstraintsChecker
	 * @throws JDOException 
	 * @throws JDOUserException if table or view is not found at DataSource metadata
	 */
	public MapRecord(TableDataSource dataSource, String tableName, FieldHelper fieldHelper, ConstraintsChecker constraintsChecker) throws JDOException {
		this(getViewDefForName(dataSource, tableName), fieldHelper, constraintsChecker);
	}
	
	/**
	 * Constructor
	 * @throws JDOException 
	 */
	public MapRecord(TableDataSource dataSource, String tableName, String... columnNames) throws JDOException {
		this(dataSource, tableName, dataSource.getFieldHelper(), null, columnNames);
	}

	/**
	 * Constructor
	 * @throws JDOException 
	 */
	public MapRecord(TableDataSource dataSource, String tableName, FieldHelper fieldHelper, String... columnNames) throws JDOException {
		this(dataSource, tableName, fieldHelper, null, columnNames);
	}

	/**
	 * Constructor
	 * @throws JDOException 
	 */
	public MapRecord(TableDataSource dataSource, String tableName, ConstraintsChecker constraintsChecker, String... columnNames) throws JDOException {
		this(dataSource, tableName, null, constraintsChecker, columnNames);
	}
	
	/**
	 * Constructor
	 * @throws JDOException 
	 */
	public MapRecord(TableDataSource dataSource, String tableName, FieldHelper fieldHelper, ConstraintsChecker constraintsChecker, String... columnNames) throws JDOException {
		super(new TableDef(tableName, filterColumns(dataSource, tableName, columnAliases(columnNames))), fieldHelper, constraintsChecker);
		values = new CaseInsensitiveValuesMap();
		columnIndexes = new HashMap<Integer, String>();
		int c = 0;
		for (String columnName : columnNames)
			columnIndexes.put(new Integer(++c), getColumnAlias(columnName));
	}
	
	private static ColumnDef[] filterColumns(TableDataSource dataSource, String tableName, String[] columnNames) {
		ColumnDef[] retval;
		ViewDef vdef = dataSource.getTableOrViewDef(tableName);
		if (null==vdef) {
			retval = new ColumnDef[columnNames.length];
			for (int c=columnNames.length-1; c>=0; c--)
				retval[c] = dataSource.createColumnDef(columnNames[c], c+1, (short) Types.NULL, null);
		} else {
			retval = vdef.filterColumns(columnAliases(columnNames));
		}
		return retval;
	}

	private static String[] columnAliases(String[] columnNames) {
		final int colCount = columnNames.length;
		String[] aliases = new String[colCount];
		for (int c=0; c<colCount; c++)
			aliases[c] = getColumnAlias(columnNames[c]);
		return aliases;
	}

	@Override
	@SuppressWarnings("unchecked")
	public Entry<String,Object>[] asEntries() {
		return values.entrySet().toArray(new Entry[values.size()]);
	}

	@Override
	public Map<String,Object> asMap() {
		return values;
	}

	@Override
	public Set<String> keySet() {
		TreeSet<String> keys = new TreeSet<String>();
		for (ColumnDef c : getTableDef().getColumns())
			keys.add(c.getName());
		return keys;
	}
	
	@Override
	public Object put(int colpos, Object value) throws ArrayIndexOutOfBoundsException {
		final String columnName = columnIndexes.get(colpos);
		if (null==columnName)
				throw new ArrayIndexOutOfBoundsException("Cannot find column at position " + String.valueOf(colpos)+" of "+String.valueOf(columnIndexes.size())+" columns");
		return values.put(columnName, value);
	}

	/**
	 * Set or replace the value of a column
	 * @param colname String
	 * @param bytes byte[]
	 * @return byte[] Previous value stored at the column
	 */
	@Override
	public Object put(String colname, byte[] bytes) {
		byte[] retval;
		Object former = values.get(colname);
		if (former==null)
			retval = null;
		else if (former instanceof byte[])
			retval = Arrays.copyOf((byte[]) former, ((byte[]) former).length);
		else 
			retval = BytesConverter.toBytes(former, Types.JAVA_OBJECT);
		values.put(colname, bytes);
		return retval;
	}

	@Override
	public Object remove(String colname) {
		Object retval = containsKey(colname) ? get(colname) : null;
		values.remove(colname);
		return retval;
	}

	@Override
	public Object remove(Object colname) {
		Object retval = containsKey(colname) ? get(colname) : null;
		values.remove(colname);
		return retval;
	}

	@Override
	public void clear() {
		values.clear();
	}

	@Override
	public boolean containsKey(Object key) {
		return values.containsKey(key);
	}

	@Override
	public boolean containsValue(Object value) {
		return values.containsValue(value);
	}

	@Override
	public Set<java.util.Map.Entry<String, Object>> entrySet() {
		HashMap<String,Object> entries = new HashMap<String,Object>();
		for (Integer columnIndex : columnIndexes.keySet()) {
			String columnName = columnIndexes.get(columnIndex);
			entries.put(columnName, values.get(columnName));
		}
		return Collections.unmodifiableSet(entries.entrySet());
	}

	@Override
	public Object apply(String key) {
		return values.get(key);
	}

	@Override
	public Object get(Object key) {
		return values.get(key);
	}

	@SuppressWarnings("unchecked")
	@Override
	public Map<String,String> getMap(String sKey) throws ClassCastException, InstantiationException, IllegalAccessException, IllegalArgumentException, NoSuchMethodException, SecurityException, ClassNotFoundException {
		return (Map<String,String>) super.getMap(sKey);
	} // getMap

	@Override
	public boolean isEmpty() {
		return values.isEmpty();
	}

	@Override
	public Object put(String key, Object value) {
		return values.put(key, value);
	}

	@Override
	public int size() {
		return values.size();
	}

	@Override
	public Collection<Object> values() {
		return values.values();
	}

	@SuppressWarnings("rawtypes")
	@Override
	public void setValue(Serializable value) throws JDOException {
		if (value instanceof Map) {
			values = new CaseInsensitiveValuesMap();
			columnIndexes = new HashMap<Integer, String>();				
			Map valmap = (Map) value;
			Iterator iter = valmap.keySet().iterator();
			int n = 0;
			while (iter.hasNext()) {
				Object key = iter.next();
				values.put(key.toString(), valmap.get(key));
				columnIndexes.put(++n, key.toString());
			}
		} else {
			throw new ClassCastException("MapRecord.setValue() Cannot cast from "+value.getClass().getName()+" to java.util.Map");
		}
	}

	@Override
	public Object getValue() throws JDOException {
		return Collections.unmodifiableMap(values);
	}

	@Override
	public void putAll(Map<? extends String,? extends Object> mValues) {
		for (Entry<? extends String,? extends Object> entry : mValues.entrySet()) {
			put(entry.getKey(), entry.getValue());
		}
	}

	private static ViewDef getViewDefForName(TableDataSource dataSource, String name) {
		final ViewDef vdef = dataSource.getTableOrViewDef(name);
		if (null==vdef) {
			if (DebugFile.trace) {
				DebugFile.writeln("JDOException at MapRecord.getViewDefForName(). Table or View not found " + name);
				DebugFile.writeln("SchemaMetadata contains " + dataSource.getMetaData().tables().size()+" tables");
				DebugFile.writeln("SchemaMetadata contains " + dataSource.getMetaData().views().size()+" views");
				StringBuilder tableList = new StringBuilder();
				tableList.append("JDC tables list = [");
				try {
					boolean first = true;
					tableList.append("SchemaMetaData tables = [");
					first = true;
					for (TableDef t : dataSource.getMetaData().tables()) {
						if (first)
							first = false;
						else
							tableList.append(",");
						tableList.append(t.getName());
					}
					tableList.append("]");
					DebugFile.writeln(tableList.toString());
					tableList.setLength(0);
					tableList.append("SchemaMetaData views = [");
					first = true;
					for (ViewDef t : dataSource.getMetaData().views()) {
						if (first)
							first = false;
						else
							tableList.append(",");
						tableList.append(t.getName());
					}
					tableList.append("]");
					DebugFile.writeln(tableList.toString());
				} catch (Exception e) {
					if (DebugFile.trace) {
						DebugFile.writeln(e.getClass().getName() + " " + e.getMessage());
					}
				}
			}
			throw new JDOUserException("MapRecord constructor. Table or view " + name + " not found");
		}
		return vdef;
	}
}

File: scala-adaptor/src/main/scala/org/judal/storage/scala/MapRecord.scala
package org.judal.storage.scala

/**
  * © Copyright 2016 the original author.
  * This file is licensed under the Apache License version 2.0.
  * You may not use this file except in compliance with the license.
  * You may obtain a copy of the License at:
  * http://www.apache.org/licenses/LICENSE-2.0
  * Unless required by applicable law or agreed to in writing,
  * software distributed under the License is distributed on an
  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  * KIND, either express or implied.
  */

import java.io.Serializable

import java.sql.Types;

import java.util.Arrays

import javax.jdo.JDOException

import org.judal.metadata.ViewDef
import org.judal.metadata.TableDef
import org.judal.metadata.ColumnDef
import org.judal.serialization.BytesConverter

import org.judal.storage.FieldHelper
import org.judal.storage.ConstraintsChecker
import org.judal.storage.EngineFactory
import org.judal.storage.table.impl.AbstractRecord
import org.judal.storage.table.TableDataSource

import scala.collection.Set
import scala.collection.mutable.DefaultEntry
import scala.collection.mutable.HashMap
import scala.collection.mutable.Map

import scala.collection.JavaConversions.asJavaCollection
import scala.collection.JavaConverters._

/**
  * Record interface implementation as a Scala mutable Map
  * @author Sergio Montoro Ten
  * @version 1.0
  */
class MapRecord(tableDefinition: ViewDef) extends AbstractRecord(tableDefinition: ViewDef) with ScalaRecord {

  /**
    * Column names are case insensitive, so specialize a HashMap to implement case insensitive lookups
    */
  class CaseInsensitiveValuesMap extends HashMap[String,AnyRef] {

    override def apply(key: String) : AnyRef = super.apply(key.toLowerCase)

    override def contains(key: String) : Boolean = super.contains(key.toLowerCase)

    override def isDefinedAt(key: String) : Boolean = super.isDefinedAt(key.toLowerCase)

    override def put (key: String, value: AnyRef) = super.put(key.toLowerCase, value)

    override def remove (key: String) = super.remove(key.toLowerCase)

    override def get (key: String) : Option[AnyRef] = super.get(key.toLowerCase)

  }

  var valuesMap = new CaseInsensitiveValuesMap
  var columnIndexes = Range(0,tableDefinition.getNumberOfColumns).map(c => c+1 -> tableDefinition.getColumns()(c).getName)(collection.breakOut): HashMap[Int,String]

  /**
    * Alternative constructor
    * @param tableDefinition ViewDef
    * @param fieldHelper FieldHelper
    * @param constraintsChecker ConstraintsChecker
    * @throws JDOException
    */
  @throws(classOf[JDOException])
  def this(tableDefinition: ViewDef, fieldHelper: FieldHelper, constraintsChecker: ConstraintsChecker) = {
    this(tableDefinition)
    setFieldHelper(fieldHelper);
    setConstraintsChecker(constraintsChecker);
  }

  /**
    * Alternative constructor
    * @param tableDefinition ViewDef
    * @param fieldHelper FieldHelper
    * @throws JDOException
    */
  @throws(classOf[JDOException])
  def this(tableDefinition: ViewDef, fieldHelper: FieldHelper) = {
    this(tableDefinition)
    setFieldHelper(fieldHelper);
  }

  /**
    * Alternative constructor
    * @param dataSource TableDataSource
    * @param tableName String
    * @throws JDOException
    */
  @throws(classOf[JDOException])
  def this(dataSource: TableDataSource, tableName: String) = {
    this(dataSource.getTableOrViewDef(tableName), dataSource.getFieldHelper)
  }

  /**
    * Alternative constructor using EngineFactory.getDefaultTableDataSource as DataSource
    * @param tableName String
    * @throws JDOException
    */
  @throws(classOf[JDOException])
  def this(tableName: String) = {
    this(EngineFactory.getDefaultTableDataSource, tableName)
  }

  /**
    * Alternative constructor
    * @param dataSource TableDataSource
    * @param tableName String
    * @param columnNames String*
    * @throws JDOException
    */
  @throws(classOf[JDOException])
  def this(dataSource: TableDataSource, tableName:String , columnNames:String*) = {
    this(new TableDef(tableName, MapRecordStatic.filterColumns(dataSource,  tableName, columnNames)), dataSource.getFieldHelper)
  }

  /**
    * Alternative constructor
    * @param tableDefinition ViewDef
    * @param constraintsChecker ConstraintsChecker
    * @throws JDOException
    */
  @throws(classOf[JDOException])
  def this(tableDefinition: ViewDef, constraintsChecker: ConstraintsChecker) = {
    this(tableDefinition)
    setConstraintsChecker(constraintsChecker);
  }

  def asEntries(): Array[java.util.Map.Entry[String,Object]] = valuesMap.asJava.entrySet.toArray(new Array[java.util.Map.Entry[String,Object]](valuesMap.size))

  def asMap(): java.util.Map[String,Object] = valuesMap.asJava

  /**
    * {@inheritDoc}
    */
  override def apply(columnName: String) : AnyRef = valuesMap.get(columnName).getOrElse(null)

  /**
    * {@inheritDoc}
    */
  override def clear() : Unit = valuesMap.clear()

  /**
    * {@inheritDoc}
    */
  override def contains(columnName: String) : Boolean = valuesMap.contains(columnName)

  /**
    * {@inheritDoc}
    */
  override def get (columnName: String) : Option[AnyRef] = valuesMap.get(columnName)

  /**
    * {@inheritDoc}
    */
  override def isDefinedAt(columnName: String) : Boolean = valuesMap.isDefinedAt(columnName)

  /**
    * {@inheritDoc}
    */
  override def keySet() : Set[String] = valuesMap.keySet

  /**
    * {@inheritDoc}
    */
  override def keys() = valuesMap.keys

  /**
    * <p>Set column value.</p>
    * @colpos int [1..columnCount()]
    * @obj AnyRef
    * @return Option[AnyRef]
    * @throws ArrayIndexOutOfBoundsException
    */
  override def put(colpos: Int, value: AnyRef) : Option[AnyRef] = {
    if (!columnIndexes.contains(colpos))
      throw new ArrayIndexOutOfBoundsException("Cannot find column at position " + colpos + " of " + columnIndexes.size +" columns")
    valuesMap.put(columnIndexes.get(colpos).get, value)
  }

  /**
    * {@inheritDoc}
    */
  override def put(colname: String, bytes: Array[Byte]) : Option[AnyRef] = {
    var retval : Array[Byte] = null
    val former = valuesMap.get(colname)
    if (former.isEmpty)
      retval = null
    else if (former.get.isInstanceOf[Array[Byte]])
      retval = Arrays.copyOf(former.asInstanceOf[Array[Byte]], former.asInstanceOf[Array[Byte]].length)
    else
      retval = BytesConverter.toBytes(former, Types.JAVA_OBJECT)
    valuesMap.put(colname, bytes);
    return Some(retval)
  }

  /**
    * {@inheritDoc}
    */
  override def remove(colname: String) : Option[AnyRef] = {
    if (valuesMap.contains(colname)) {
      val retval = valuesMap.get(colname)
      valuesMap.remove(colname)
      retval
    }  else {
      valuesMap.remove(colname)
      None
    }
  }

  /**
    * {@inheritDoc}
    */
  override def isEmpty() : Boolean = valuesMap.isEmpty

  /**
    * {@inheritDoc}
    */
  override def size() : Int = valuesMap.size

  /**
    * {@inheritDoc}
    */
  override def values() : Iterable[AnyRef] = valuesMap.values

  /**
    * {@inheritDoc}
    */
  override def getMap(key: String) : Map[String,String] = super.getMap(key).asInstanceOf[Map[String,String]]

  /**
    * {@inheritDoc}
    */
  @throws(classOf[JDOException])
  override def setValue(value: Serializable): Unit = {
    if (value.isInstanceOf[Map[String,AnyRef] @unchecked]) {
      valuesMap = new CaseInsensitiveValuesMap()
      columnIndexes = new HashMap[Int, String]()
      val valmap = value.asInstanceOf[Map[String,AnyRef]]
      val iter = valmap.keysIterator
      var n = 0
      while (iter.hasNext) {
        n += 1
        val key = iter.next
        valuesMap.put(key.asInstanceOf[String], valmap.get(key))
        columnIndexes.put(n, key.asInstanceOf[String])
      }
    } else {
      throw new ClassCastException("MapRecord.setValue() Cannot cast from "+value.getClass().getName()+" to java.util.Map");
    }
  }

  /**
    * {@inheritDoc}
    */
  @throws(classOf[JDOException])
  override def getValue() : scala.collection.mutable.Map[String,AnyRef] = valuesMap

  /**
    * <p>Set value at internal collection</p>
    * @param sKey Column Name
    * @param oObj Field Value
    * @throws NullPointerException If sKey is <b>null</b>
    * @throws IllegalArgumentException If no column with given name is not found at record columns list
    */
  override def put(colname: String, value: AnyRef): Option[AnyRef] = {
    val retval = valuesMap.get(colname)
    valuesMap.put(colname, value)
    retval
  }

  /**
    * {@inheritDoc}
    */
  override def +=(kv : (String, AnyRef)) = {
    valuesMap += kv
    this
  }

  /**
    * {@inheritDoc}
    */
  override def update(columnName: String, columnValue: AnyRef) : Unit = {
    valuesMap.update(columnName, columnValue)
  }

  /**
    * {@inheritDoc}
    */
  override def -=(colname: String) = {
    valuesMap.remove(colname)
    this
  }

  /**
    * {@inheritDoc}
    */
  def iterator: Iterator[(String, AnyRef)] = valuesMap.iterator

}

private object MapRecordStatic {

  def filterColumns(dataSource: TableDataSource, tableName: String , columnNames: Seq[String]) = {
    var retval : Array[ColumnDef] = null
    val vdef = dataSource.getTableOrViewDef(tableName)
    if (null==vdef) {
      retval = new Array[ColumnDef](columnNames.length)
      for (c <- 0 to columnNames.length-1)
        retval(c) = dataSource.createColumnDef(columnNames(c), c+1, Types.NULL, null)
    } else {
      retval = vdef.filterColumns(columnAliases(columnNames))
    }
    asJavaCollection(retval.toSeq)
  }

  def columnAliases(columnNames: Seq[String]) : Array[String] = {
    val colCount = columnNames.length
    val aliases = new Array[String](colCount)
    for (c <- 0 to colCount-1)
      aliases(c) = getColumnAlias(columnNames(c))
    return aliases;
  }

  def getColumnAlias(columnName: String) : String = {
    for (n <- 0 to columnName.length-6)
      if (columnName.charAt(n)==' ' &&
        (columnName.charAt(n+1)=='A' || columnName.charAt(n+1)=='a') &&
        (columnName.charAt(n+2)=='S' || columnName.charAt(n+2)=='s') &&
        columnName.charAt(n+3)==' ')
        return columnName.substring(n+4)
    return columnName
  }

}